rfcs/20180905-deprecate-collections.md

```diff
@@ -0,0 +1,612 @@
+   <td>Not used much
+   </td>
+   <td>
+   </td>
+  </tr>
+  <tr>
+   <td>Updates
+   </td>
+   <td>UPDATE_OPS
+   </td>
+   <td>Model.updates
+   </td>
+   <td>BatchNormalization.updates
+   </td>
+  </tr>
+  <tr>
+   <td>Iterators
+   </td>
+   <td>GLOBAL_ITERATORS
+   </td>
+   <td>Object based checkpointing handles saveables generically
+   </td>
+   <td>
+   </td>
+  </tr>
+  <tr>
+   <td>SavedModel
+   </td>
+   <td>INIT_OP
+   </td>
+   <td>Keep
+   </td>
+   <td>
+   </td>
+  </tr>
+  <tr>
+   <td>SavedModel
+   </td>
+   <td>LOCAL_INIT_OP
+   </td>
+   <td>Keep
+   </td>
+   <td>
+   </td>
+  </tr>
+  <tr>
+   <td>SavedModel
+   </td>
+   <td>READY_OP
+   </td>
+   <td>Keep
+   </td>
+   <td>
+   </td>
+  </tr>
+  <tr>
+   <td>SavedModel
+   </td>
+   <td>READY_FOR_LOCAL_INIT_OP
+   </td>
+   <td>Keep
+   </td>
+   <td>
+   </td>
+  </tr>
+  <tr>
+   <td>SavedModel
+   </td>
+   <td>GLOBAL_STEP
+   </td>
+   <td>Keep
+   </td>
+   <td>
+   </td>
+  </tr>
+</table>
+
+
+For some specific use cases, we'll elaborate how we will handle them
+
+
+### **Global variables, trainable variables**
+
+For the keras models use case, Model.variables already tracks all variables created in a Keras model. Similarly, for trainable variables, Model.trainable_variables tracks all trainable variables. 
+
+In case of custom graph code, variable_creator_scope can be used to collect up variables.
+
+Example code might look like
+
+```python
+class VariableTracker(object):
+  def __init__(self):
+    self.variables = []
+
+  def variable_tracker(self, next_creator, **kwargs):
+    v = next_creator(**kwargs)
+    self.variables.append(v)
+    return v
+
+VariableTracker tracker
+with tf.variable_creator_scope(tracker.variable_tracker):
+  ...
+  a = tf.Variable(0)
+  ...
+
+assert tracker.variables == [a]
+```
+
+We might want to expose VariableTracker as well.
+
+
+### **Local variables**
+
+Local variables are primarily used for metrics. With the move to objects for metrics, one can access the variables created for the metrics via the Metric object. 
+
+
+### **Moving average variables**
+
+Collections are used in two places here. We'll remove these calls and ask callers to explicitly pass in a variable list 
+
+
+
+*   [MovingAverageOptimizer.swapping_saver](https://github.com/tensorflow/tensorflow/blob/r1.10/tensorflow/contrib/opt/python/training/moving_average_optimizer.py#L140)
+*   [ExponentialMovingAverage.variables_to_restore](https://github.com/tensorflow/tensorflow/blob/r1.10/tensorflow/python/training/moving_averages.py#L473). We'll have to expose the variables created by ExponentialMovingAverage as they would have to be passed in here for this method.
+
+
+### **Tables**
+
+
+#### **Keras models**
+
+We'll provide a `track_table` method in the backend (similar to `track_variable`) and we'll make sure that when model.fit is called, tables are initialized similar to the way we initialize variables. Changes we'll have to make here are the following
```

Comment:
`track_variable` and `track_table` seem suspiciously similar to the global behavior we're trying to remove.  

Are these okay because they are special purpose?  Is there a rule of thumb for when global == OK?

Reply:
yeah they are indeed. The thinking is to provide them as a backstop because forcing everyone to migrate to managing their state immediately would be tough. As a principle, we will try not to use them in the tensorflow code base at all.

Reply:
Tables in particular seem too special case. If we are just capturing initializers, maybe this can be generalized for that? Variables and initializers as special cases; everything else we keep out of global scope? We can handle assets through the EstimatorSpec (and the same for similar entities, perhaps?)